### PR TITLE
Fix PickLoadSigns on SignExt feature instructions

### PR DESCRIPTION
**Pull request by kripken**
I believe the history here is that

1. We added a `PickLoadSigns` pass. It checks if a load from memory is stored in
a local that is only every used in a signed or an unsigned manner. If it is, we can
adjust the sign of the load (`load8_u/s`) to do the sign/unsign during the load.
2. The pass finds each LocalGet and looks either 2 or 3 parents above it. For
a sign operation, we need to look up 3, since the operation is `x << K >> K`. For
an unsigned, we need only 2, since we have `x & M`. We hardcoded those
numbers 2 and 3.
3. We added the SignExt feature, which adds `i32.extend8_s`. This does a sign
extend with a single instruction, not two nested ones, so now we can sign-
extend at depth 2, unlike before. `Properties::getSignExtValue` was updated
for this, but not the pass `PickLoadSigns`.

The bug that is fixed here is that we looked at depth 3 for a sign-extend, and
we blindly accepted it if we found one. So we ended up accepting
`(i32.extend8_s (ANYTHING (x)))`, which is a sign-extend of something, but
not of `x`, which is bad.

We were also missing an optimization opportunity, as we didn't look for
depth 2 sign extends.

This bug is quite old, from when Properties got SignExt support, in #3910.
But the blame isn't there - to notice this then, we'd have had to check each
caller of `getSignExtValue` throughout the codebase, which isn't reasonable.
The fault is mine, from the first write-up of `PickLoadSigns` in 2017: the code
should have been fully general, handling 2/3 and checking the output when
it does so (adding `== curr`, that the sign/zero-extended value is the one we
expect). That is what this PR does.

### CODE REVIEW

**Comment by kripken** [comment]
(diff without whitespace is a bit smaller)

File: test/lit/passes/pick-load-signs_sign-ext.wast [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+ ;; CHECK:      (func $load-valid (type $1)
+ ;; CHECK-NEXT:  (local $temp i32)
+ ;; CHECK-NEXT:  (local.set $temp
+ ;; CHECK-NEXT:   (i32.load8_s
+ ;; CHECK-NEXT:    (i32.const 22)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (drop
+ ;; CHECK-NEXT:   (i32.extend8_s
```
Comment (tlively) [inline test/lit/passes/pick-load-signs_sign-ext.wast:59]: Is there another pass that will be able to optimize out the `i32.extend8_s`?

**Review by tlively — APPROVED** [review]

File: test/lit/passes/pick-load-signs_sign-ext.wast (hunk 1 above)
Comment (kripken) [reply test/lit/passes/pick-load-signs_sign-ext.wast:59]: That should happen over here:

https://github.com/WebAssembly/binaryen/blob/9e11c5fd12a7c6392ac62979cc16c62a368f6e33/src/passes/OptimizeInstructions.cpp#L588-L594

I haven't verified yet if it does. If not, I'll open a PR.

Comment (kripken) [reply test/lit/passes/pick-load-signs_sign-ext.wast:59]: Ah, that code is in visitBinary... but the signExt operations are Unary. I opened https://github.com/WebAssembly/binaryen/pull/7072

**Pull request merged**